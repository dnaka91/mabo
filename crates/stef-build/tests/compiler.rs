use std::{
    fs,
    path::{Path, PathBuf},
};

use insta::{assert_snapshot, glob, with_settings};
use stef_build::Opts;
use stef_parser::Schema;

fn strip_path(path: &Path) -> PathBuf {
    path.strip_prefix(concat!(env!("CARGO_MANIFEST_DIR"), "/tests/inputs"))
        .or_else(|_| path.strip_prefix(concat!(env!("CARGO_MANIFEST_DIR"), "/tests/inputs_extra")))
        .unwrap()
        .to_owned()
}

#[test]
fn compile_schema() {
    glob!("inputs/*.stef", |path| {
        let input = fs::read_to_string(path).unwrap();
        let value = Schema::parse(input.as_str(), Some(&strip_path(path))).unwrap();
        let value = stef_compiler::simplify_schema(&value);
        let value = stef_build::compile_schema(&Opts::default(), &value);
        let value = prettyplease::unparse(&syn::parse2(value.clone()).unwrap());

        with_settings!({
            description => input.trim(),
            omit_expression => true,
        }, {
            assert_snapshot!("compile", value);
        });
    });
}

#[test]
fn compile_schema_extra() {
    glob!("inputs_extra/*.stef", |path| {
        let input = fs::read_to_string(path).unwrap();
        let value = Schema::parse(input.as_str(), Some(&strip_path(path))).unwrap();
        let value = stef_compiler::simplify_schema(&value);
        let value = stef_build::compile_schema(&Opts::default(), &value);
        let value = prettyplease::unparse(&syn::parse2(value.clone()).unwrap());

        with_settings!({
            description => input.trim(),
            omit_expression => true,
        }, {
            assert_snapshot!("compile_extra", value);
        });
    });
}
