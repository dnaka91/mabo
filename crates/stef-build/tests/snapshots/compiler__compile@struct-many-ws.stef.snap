---
source: crates/stef-build/tests/compiler.rs
expression: "/// Some comment\n       struct      Sample<\n            A,\n            B\n        >  {\n\n      a:    u32    @1,\n      b:    bool   @2,\n\n    }"
input_file: crates/stef-parser/tests/inputs/struct-many-ws.stef
---
#[allow(unused_imports)]
use ::stef::buf::{Decode, Encode};
/// Some comment
#[derive(Clone, Debug, PartialEq)]
#[allow(clippy::module_name_repetitions, clippy::option_option)]
pub struct Sample<A, B> {
    pub a: u32,
    pub b: bool,
}
#[automatically_derived]
impl<A, B> ::stef::Encode for Sample<A, B>
where
    A: ::stef::buf::Encode,
    B: ::stef::buf::Encode,
{
    #[allow(
        clippy::borrow_deref_ref,
        clippy::explicit_auto_deref,
        clippy::needless_borrow,
        clippy::too_many_lines,
    )]
    fn encode(&self, w: &mut impl ::stef::BufMut) {
        ::stef::buf::encode_field(
            w,
            1,
            |w| {
                ::stef::buf::encode_u32(w, self.a);
            },
        );
        ::stef::buf::encode_field(
            w,
            2,
            |w| {
                ::stef::buf::encode_bool(w, self.b);
            },
        );
        ::stef::buf::encode_u32(w, ::stef::buf::END_MARKER);
    }
}
#[automatically_derived]
impl<A, B> ::stef::Decode for Sample<A, B>
where
    A: ::std::fmt::Debug + ::stef::buf::Decode,
    B: ::std::fmt::Debug + ::stef::buf::Decode,
{
    #[allow(clippy::type_complexity, clippy::too_many_lines)]
    fn decode(r: &mut impl ::stef::Buf) -> ::stef::buf::Result<Self> {
        let mut a: Option<u32> = None;
        let mut b: Option<bool> = None;
        loop {
            match ::stef::buf::decode_id(r)? {
                ::stef::buf::END_MARKER => break,
                1 => a = Some(::stef::buf::decode_u32(r)?),
                2 => b = Some(::stef::buf::decode_bool(r)?),
                _ => continue,
            }
        }
        Ok(Self {
            a: a
                .ok_or(::stef::buf::Error::MissingField {
                    id: 1,
                    name: Some("a"),
                })?,
            b: b
                .ok_or(::stef::buf::Error::MissingField {
                    id: 2,
                    name: Some("b"),
                })?,
        })
    }
}

